File: elixir/2015/day02b/Program.exs
#!/usr/bin/env elixir

defmodule Program do
  def main(args) do
    case args do
      [filename] ->
        case process(filename) do
          {:ok, result} ->
            IO.puts("result = #{result}")
          {:error, reason} ->
            IO.puts(:stderr, "error while processing file `#{filename}': #{reason}")
            System.halt(1)
        end
      _ ->
        usage()
    end
  end

  defp usage do
    IO.puts(:stderr, "usage: Program.exs <input file>")
    System.halt(1)
  end

  defp process(filename) do
    case File.read(filename) do
      {:ok, content} ->
        total_len =
          content
          |> String.split()
          |> Enum.reduce(0, fn
            line, acc ->
              parts = String.split(line, "x")
              dims = Enum.map(parts, fn x -> String.to_integer(x) end)
              l = Enum.at(dims, 0)
              w = Enum.at(dims, 1)
              h = Enum.at(dims, 2)
              perim1 = 2 * (l + w)
              perim2 = 2 * (l + h)
              perim3 = 2 * (w + h)
              present_len = min(perim1, min(perim2, perim3))
              bow_len = l * w * h
              acc + present_len + bow_len
          end)
        {:ok, total_len}
      {:error, reason} ->
        {:error, reason}
    end
  end
end

Program.main(System.argv())

File: erlang/2015/day02b/Program.erl
#!/usr/bin/env escript

main(Args) ->
  case Args of
    [Filename] ->
      case process(Filename) of
        {ok, Result} ->
          io:format("result = ~p~n", [Result]);
        {error, Reason} ->
          io:format(standard_error, "error while processing file `~s': ~p~n", [Filename, Reason]),
          halt(1)
      end;
    _ ->
      usage()
  end.

usage() ->
  io:format(standard_error, "usage: Program.erl <input file>~n", []),
  halt(1).

process(Filename) ->
  case file:read_file(Filename) of
    {ok, Chars} ->
      Lines = string:lexemes(Chars, "\n"),
      Result = lists:foldl(
        fun
          (Line, Acc) ->
            case string:lexemes(Line, "x") of
              [LStr, WStr, HStr] ->
                {L, _} = string:to_integer(LStr),
		{W, _} = string:to_integer(WStr),
                {H, _} = string:to_integer(HStr),
                Perim1 = 2 * (L + W),
                Perim2 = 2 * (L + H),
                Perim3 = 2 * (W + H),
                PresentLen = lists:min([Perim1, Perim2, Perim3]),
                BowLen = L * W * H,
                Acc + PresentLen + BowLen;
              _ ->
                Acc
            end
        end,
        0,
        Lines),
      {ok, Result};
    {error, Reason} ->
      {error, Reason}
  end.
